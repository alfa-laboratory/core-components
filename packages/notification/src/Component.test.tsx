import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Button } from '@alfalab/core-components-button';

import { Notification } from './index';

jest.useFakeTimers();

describe('Notification', () => {
    describe('Snapshots tests', () => {
        const onClose = jest.fn();

        it('should match snapshot', () => {
            const { baseElement } = render(
                <Notification badge='positive' title='title' onClose={onClose}>
                    text
                </Notification>,
            );

            expect(baseElement).toMatchSnapshot();
        });
    });

    it('should set `data-test-id` attribute', () => {
        const dataTestId = 'test-id';
        const { getByTestId } = render(<Notification dataTestId={dataTestId} />);

        expect(getByTestId(dataTestId).tagName).toBe('DIV');
    });

    it('should forward ref', () => {
        const ref = jest.fn();
        const dataTestId = 'test-id';
        const { getByTestId } = render(<Notification ref={ref} dataTestId={dataTestId} />);

        expect(ref.mock.calls).toEqual([[getByTestId(dataTestId)]]);
    });

    describe('Classes tests', () => {
        it('should set `className` class', () => {
            const className = 'test-class';
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Notification className={className} dataTestId={dataTestId} />,
            );

            const el = getByTestId(dataTestId);

            expect(el).toHaveClass(className);
        });

        it('should set `visible` class', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Notification visible={true} dataTestId={dataTestId} />);

            const el = getByTestId(dataTestId);

            expect(el).toHaveClass('isVisible');
        });
    });

    describe('Callbacks tests', () => {
        it('should call `onClose` prop', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Notification hasCloser={true} onClose={cb} dataTestId={dataTestId} />,
            );

            const el = getByTestId(dataTestId);
            const closeEl = el.querySelector('[aria-label="закрыть"]') as Element;

            fireEvent.click(closeEl);

            expect(cb).toBeCalledTimes(1);
        });

        it('should call `onCloseTimeout` prop', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Notification
                    autoCloseDelay={100}
                    onCloseTimeout={cb}
                    visible={true}
                    dataTestId={dataTestId}
                />,
            );

            const el = getByTestId(dataTestId);

            fireEvent.mouseEnter(el);
            fireEvent.mouseLeave(el);

            jest.advanceTimersByTime(100);

            expect(cb).toBeCalledTimes(1);
        });

        it('should call `onMouseEnter` prop', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Notification onMouseEnter={cb} dataTestId={dataTestId} />,
            );

            const el = getByTestId(dataTestId);

            fireEvent.mouseEnter(el);

            expect(cb).toBeCalledTimes(1);
        });

        it('should call `onMouseLeave` prop', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Notification onMouseLeave={cb} dataTestId={dataTestId} />,
            );

            const el = getByTestId(dataTestId);

            fireEvent.mouseLeave(el);

            expect(cb).toBeCalledTimes(1);
        });

        it('should call `onClickOutside` prop', async () => {
            const cb = jest.fn();
            const dataTestId = 'btn-test-id';
            const { getByTestId } = render(
                <div>
                    <Button dataTestId={dataTestId}>btn</Button>
                    <Notification onClickOutside={cb} visible={true} />
                </div>,
            );

            const el = getByTestId(dataTestId);

            fireEvent.click(el);

            expect(cb).toBeCalledTimes(1);
        });

        it('should not call `onClickOutside` if clicked inside', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Notification onClickOutside={cb} visible={true} dataTestId={dataTestId} />,
            );

            const el = getByTestId(dataTestId);

            fireEvent.click(el);
            fireEvent.click(el.firstElementChild as HTMLElement);

            expect(cb).toBeCalledTimes(0);
        });

        it('should not call `onClickOutside` if clicked inside another', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <div>
                    <Notification visible={true} dataTestId={dataTestId} />
                    <Notification onClickOutside={cb} visible={true} />
                </div>,
            );

            const el = getByTestId(dataTestId);

            fireEvent.click(el);
            fireEvent.click(el.firstElementChild as HTMLElement);

            expect(cb).toBeCalledTimes(0);
        });
    });

    it('should unmount without errors', () => {
        const { unmount } = render(<Notification />);

        expect(unmount).not.toThrowError();
    });
});
