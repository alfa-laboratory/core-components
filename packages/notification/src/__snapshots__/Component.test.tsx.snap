// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Snapshots tests should match snapshot 1`] = `
<body>
  <div />
  <div
    alfa-portal-container=""
  >
    <div>
      <div
        class="component hasCloser notificationComponent"
        style="top: 108px; z-index: 1000;"
      >
        <div
          class="contentWrap"
        >
          <div
            class="toastContent content"
          >
            <div
              class="leftAddons"
            >
              <div
                class="component m icon positive badge"
                data-test-id="badge"
              >
                <svg
                  class="badgeIcon"
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    clip-rule="evenodd"
                    d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10zm-1.5-6L17 9.5 15.5 8l-5 5-2-2L7 12.5l3.5 3.5z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </div>
            <div>
              <div
                class="title"
              >
                title
              </div>
              <div
                class="children"
              >
                text
              </div>
            </div>
          </div>
          <button
            aria-label="закрыть"
            class="component ghost m component ghost iconOnly closeButton"
            type="button"
          >
            <span
              class="addons"
            >
              <svg
                fill="currentColor"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M10.586 12L6 16.5 7.5 18l4.5-4.586L16.5 18l1.5-1.5-4.586-4.5L18 7.5 16.5 6 12 10.586 7.5 6 6 7.5l4.586 4.5z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
